source/openqasm/ast.py

```diff
@@ -0,0 +1,1010 @@
+        angle[32]
+    """
+
+    type: SingleDesignatorTypeName
+    designator: Optional[Expression]
+
+
+class SingleDesignatorTypeName(Enum):
+    """
+    Single designator types
+    """
+
+    int = auto()
+    uint = auto()
+    float = auto()
+    angle = auto()
+
+
+@dataclass
+class BitType(ClassicalType):
+    """
+    Bit type
+
+    Example::
+
+        bit[8]
+        creg[8]
+    """
+
+    type: BitTypeName
+    designator: Optional[Expression]
+
+
+class BitTypeName(Enum):
+    """
+    Bit types
+    """
+
+    bit = auto()
+    creg = auto()
+
+
+@dataclass
+class NoDesignatorType(ClassicalType):
+    """
+    Type with no designators.
+
+    Example::
+
+        bool
+        duration
+        stretch
+    """
+
+    type: NoDesignatorTypeName
+
+
+class NoDesignatorTypeName(Enum):
+    """
+    No designator type names
+    """
+
+    bool = auto()
+    duration = auto()
+    stretch = auto()
+
+
+@dataclass
+class ComplexType(ClassicalType):
+    """
+    Complex Type. Its real and imaginary parts are based on other classical types.
+
+    Example::
+
+        complex[int[32]]
+        complex[float[32]]
+    """
+
+    base_type: Union[SingleDesignatorType, DoubleDesignatorType]
+
+
+class IndexIdentifier(OpenNode):
+    """
+    Quantum or classical identifier,
+    indexed or not indexed.
+
+    #TODO: we will have to update IndexExpression,
+    then IndexIdentifier will be subclass of Expression.
+
+    Example::
+
+        q
+        b
+        b[1]
+        b[3:5]
+        b || c
+    """
+
+
+@dataclass
+class Subscript(IndexIdentifier):
+    """
+    Indexed identifier with an integer
+    as subscript.
+
+    Example:
+
+        segments[0]
+        qubits[1]
+    """
+
+    name: str
+    index: Expression
+
+
+@dataclass
+class Selection(IndexIdentifier):
+    """
+    Indexed identifier with multiple integers
+    as subscript.
+
+    Example::
+
+        segments[1, 2]
+        qubits[1, 2, 3]
+    """
+
+    name: str
+    indices: List[Expression]
+
+
+@dataclass
+class Slice(IndexIdentifier):
+    """
+    Indexed identifier with a range
+    as subscript.
+
+    Example::
+
+        segments[1:2]
+        qubits[1:10:1]
+    """
+
+    name: str
+    range: RangeDefinition
+
+
+@dataclass
+class RangeDefinition(OpenNode):
+    """
+    Range definition.
+
+    Example::
+
+        [1:2]
+        [1:1:10]
+        [:]
+    """
+
+    start: Optional[Expression]
+    end: Optional[Expression]
+    step: Optional[Expression]
+
+
+@dataclass
+class Concatenation(IndexIdentifier):
+    """
+    Combination of two arrays.
+
+    Example::
+
+        segment1 || segment2
+        qubits[0:10] || qubits[15:20]
+    """
+
+    lhs: Union[IndexIdentifier, Identifier]
+    rhs: Union[IndexIdentifier, Identifier]
+
+
+@dataclass
+class CalibrationGrammarDeclaration(Statement):
+    """
+    Calibration grammar declaration
+
+    Example::
+
+        defcalgrammar "openpulse";
+    """
+
+    calibration_grammar: str
```

Comment:
should this `str` be `StringLiteral`?

Reply:
`StringLiteral` node is just a wrapper over a `str` value, but it gives us a span. I am in favor of StringLiteral.